File: projects/TsProject/src/common/observable.ts
export class Observable<T, TData = never> {
    public get closed() { return this._closed; }
    public get count() { return this._closed ? 0 : this._subscribers.length; }
    private _closed: boolean = false;
    private _subscribers: Array<Subscriber<T>> = [];
    private _defaultValue: TData;
    private _subscribe?: (subscriber: Subscriber<T>, defaultValue?: TData) => void;

    constructor(subscribe?: (subscriber: Subscriber<T>, defaultValue?: TData) => void) {
        this._subscribe = subscribe;
    }
    public subscribe(callback: (val: T) => void, immediately: boolean = true): Subscription {
        if (this._closed)
            return undefined;
        let subscriber = new Subscriber<T>(callback);
        this._subscribers.push(subscriber);
        if (immediately) {
            this._subscribe?.call(this, subscriber, this._defaultValue);
        }
        return subscriber;
    }
    public next(value: T) {
        if (this._closed) return;
        let closed = 0;
        this._subscribers.forEach(subscriber => {
            if (subscriber.closed) closed++;
            else subscriber.next(value);
        });
        if (closed > 0) {
            this._subscribers = this._subscribers.filter(subscriber => !subscriber.closed);
        }
    }
    public unsubscribe() {
        if (this._closed) return;
        this._subscribers.forEach(subscriber => subscriber.unsubscribe());
        this._subscribers = [];
    }
    public dispose() {
        if (this._closed) return;
        this.unsubscribe();
        this._closed = true;
    }

    public setDefault(data: TData) {
        this._defaultValue = data;
    }
}
export class Subscription {
    public get closed() { return this._closed; }
    private _closed: boolean = false;
    public unsubscribe() {
        this._closed = true;
    }
}
export class Subscriber<T> extends Subscription {
    private _closedFn: () => void;
    private _callback: (val: T) => void;

    constructor(callback: (val: T) => void) {
        super();
        this._callback = callback;
    }
    public unsubscribe() {
        if (this.closed) return;
        super.unsubscribe();

        let cb = this._closedFn;
        this._callback = undefined;
        this._closedFn = undefined;

        if (cb) cb();
    }
    public onClosed(fn: () => void) {
        this._closedFn = fn;
    }
    public next(value: T) {
        if (this.closed) return;
        this._callback?.call(this, value);
    }
}

File: projects/Assets/XOR/Runtime/Src/Common/Observable.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

namespace XOR
{
    public class Subscription
    {
        public bool Closed { get; protected set; }

        public virtual void Unsubscribe()
        {
            this.Closed = true;
        }
    }

    public class Subscriber : Subscription
    {
        private Action _close;
        private Action _callback;

        public Subscriber(Action callback)
        {
            this._callback = callback;
        }
        public override void Unsubscribe()
        {
            if (this.Closed) return;
            base.Unsubscribe();

            Action fn = this._close;
            this._close = null;
            this._callback = null;

            fn?.Invoke();
        }
        public void OnClose(Action handler)
        {
            this._close = handler;
        }

        public void Next()
        {
            if (this.Closed) return;
            this._callback?.Invoke();
        }
    }
    public class Subscriber<T> : Subscription
    {
        private Action _close;
        private Action<T> _callback;

        public Subscriber(Action<T> callback)
        {
            this._callback = callback;
        }

        public override void Unsubscribe()
        {
            if (this.Closed) return;
            base.Unsubscribe();

            Action fn = this._close;
            this._close = null;
            this._callback = null;

            fn?.Invoke();
        }
        public void OnClose(Action handler)
        {
            this._close = handler;
        }

        public void Next(T value)
        {
            if (this.Closed) return;
            this._callback?.Invoke(value);
        }
    }

    public class Observable
    {
        public bool Closed { get; protected set; }
        public int Count { get => this.subscribers != null ? this.subscribers.Count : 0; }

        protected Action<Subscriber> subscribe;
        protected HashSet<Subscriber> subscribers;

        public Observable() : this(null) { }
        public Observable(Action<Subscriber> subscribe)
        {
            this.subscribe = subscribe;
            this.subscribers = new HashSet<Subscriber>();
        }

        public virtual Subscription Subscribe(Action callback, bool immediately = true)
        {
            if (this.Closed)
                return null;
            Subscriber subscriber = new Subscriber(callback);
            this.subscribers.Add(subscriber);
            if (immediately)
            {
                this.subscribe?.Invoke(subscriber);
            }
            return subscriber;
        }
        public void Next()
        {
            if (this.Closed) return;

            int closed = 0;
            foreach (var subscriber in this.subscribers)
            {
                if (subscriber.Closed)
                {
                    closed++;
                    continue;
                }
                subscriber.Next();
            }

            if (closed > 0)
            {
                foreach (var subscriber in this.subscribers.Where(s => s.Closed))
                {
                    this.subscribers.Remove(subscriber);
                }
            }
        }
        public void Unsubscribe()
        {
            if (this.Closed) return;
            foreach (var subscriber in this.subscribers)
            {
                subscriber.Unsubscribe();
            }
            this.subscribers.Clear();
        }
        public void Dispose()
        {
            if (this.Closed) return;
            this.Closed = true;
            this.Unsubscribe();
        }
    }
    public class Observable<T>
    {
        public bool Closed { get; protected set; }
        public int Count { get => this.subscribers != null ? this.subscribers.Count : 0; }

        protected Action<Subscriber<T>> subscribe;
        protected HashSet<Subscriber<T>> subscribers;

        public Observable() : this(null) { }
        public Observable(Action<Subscriber<T>> subscribe)
        {
            this.subscribe = subscribe;
            this.subscribers = new HashSet<Subscriber<T>>();
        }

        public virtual Subscription Subscribe(Action<T> callback, bool immediately = true)
        {
            if (this.Closed)
                return null;
            Subscriber<T> subscriber = new Subscriber<T>(callback);
            this.subscribers.Add(subscriber);
            if (immediately)
            {
                this.subscribe?.Invoke(subscriber);
            }
            return subscriber;
        }
        public void Next(T value)
        {
            if (this.Closed) return;

            int closed = 0;
            foreach (var subscriber in this.subscribers)
            {
                if (subscriber.Closed)
                {
                    closed++;
                    continue;
                }
                subscriber.Next(value);
            }

            if (closed > 0)
            {
                foreach (var subscriber in this.subscribers.Where(s => s.Closed))
                {
                    this.subscribers.Remove(subscriber);
                }
            }
        }
        public void Unsubscribe()
        {
            if (this.Closed) return;
            foreach (var subscriber in this.subscribers)
            {
                subscriber.Unsubscribe();
            }
            this.subscribers.Clear();
        }
        public void Dispose()
        {
            if (this.Closed) return;
            this.Closed = true;
            this.Unsubscribe();
        }
    }
    public class Observable<T, TState> : Observable<T>
    {
        protected Action<Subscriber<T>, TState> subscribe2;
        protected TState state;

        public Observable(Action<Subscriber<T>, TState> subscribe)
        {
            this.subscribe2 = subscribe;
        }
        public override Subscription Subscribe(Action<T> callback, bool immediately = true)
        {
            if (this.Closed)
                return null;
            if (this.subscribe2 != null && immediately)
            {
                Subscriber<T> subscriber = new Subscriber<T>(callback);
                this.subscribers.Add(subscriber);
                this.subscribe2(subscriber, state);
                return subscriber;
            }
            else
            {
                return base.Subscribe(callback, immediately);
            }
        }
        public void SetDefault(TState state)
        {
            this.state = state;
        }
    }
}
